### ISSUE TITLE
a couple of problems with examples

### ISSUE THREAD

**Issue by refried** [http4s/http4s#178]
Sorry, this ticket is kind of a mess, but I wanted to record it before I forgot.

1) When I tried ran the BlazeExample, I ran into trouble after clicking the Redirect example — it redirected me from http://localhost:8080/http4s/redirect to http://localhost:8080/http4s .  From there, the rest of the relative links from html.index() won't work; e.g. the relative link to "ping" resolves to http://localhost:8080/ping, rather than http://localhost:8080/http4s/ping  

I didn't make a PR because I wasn't sure what the best solution was, or whether it needed to be mirrored in some other example services, but my local diff is below.

2) The "push" example uses `org.http4s.scalaxml.ElemInstances#xmlEnocder` which defaults to `application/xml`.  My browser (probably Chrome, I forget, it was a while ago) renders this differently from `text/html`; specifically, it doesn't render inline images, which I figured defeats the purpose of the "push" example.  I overrode the content type as seen below, but wasn't sure if that was the most elegant fix.

Thanks again, you folks are making http4s awesome.
```
diff --git a/examples/blaze/src/main/scala/com/example/http4s/blaze/BlazeExample.scala b/examples/blaze/src/main/scala/com/example/http4s/blaze/BlazeExample.scala
index e91812f..b6cc779 100644
--- a/examples/blaze/src/main/scala/com/example/http4s/blaze/BlazeExample.scala
+++ b/examples/blaze/src/main/scala/com/example/http4s/blaze/BlazeExample.scala
@@ -7,7 +7,7 @@ import org.http4s.server.blaze.BlazeBuilder
 
 object BlazeExample extends App {
   BlazeBuilder.bindHttp(8080)
-    .mountService(ExampleService.service, "/http4s")
+    .mountService(ExampleService.service, "/http4s/")
     .run
     .awaitShutdown()
 }
diff --git a/examples/src/main/scala/com/example/http4s/ExampleService.scala b/examples/src/main/scala/com/example/http4s/ExampleService.scala
index 66153b8..0ef8934 100644
--- a/examples/src/main/scala/com/example/http4s/ExampleService.scala
+++ b/examples/src/main/scala/com/example/http4s/ExampleService.scala
@@ -54,7 +54,7 @@ object ExampleService {
 
     case req @ GET -> Root / "redirect" =>
       // Not every response must be Ok using a EntityEncoder: some have meaning only for specific types
-      TemporaryRedirect(uri("/http4s"))
+      TemporaryRedirect(uri("/http4s/"))
 
     case GET -> Root / "content-change" =>
       // EntityEncoder typically deals with appropriate headers, but they can be overridden
@@ -113,7 +113,7 @@ object ExampleService {
     case req @ GET -> Root / "push" =>
       // http4s intends to be a forward looking library made with http2.0 in mind
       val data = <html><body><img src="image.jpg"/></body></html>
-      Ok(data).push("/image.jpg")(req)
+      Ok(data).withHeaders(`Content-Type`(`text/html`)).push("image.jpg")(req)
 
     case req @ GET -> Root / "image.jpg" =>
       StaticFile.fromResource("/nasa_blackhole_image.jpg", Some(req))
```

**Comment by rossabaker** [comment]
Good catches.  This would be welcome as a pull request.  I understand the TemporaryRedirect, but I need to mull over what the trailing slash means in mounting a service in Blaze.  We should put some thought into ensuring BlazeBuilder and the servlet builders act similarly.

We recently changed XML literals' default content type from `text/html` to `application/xml`, figuring that scala.xml will more commonly be used for XML than inline HTML templating.  The opposite was true in our example service.  `scala.xml.Elem` is a tricky one, because it a few reasonable default content types. Your solution, or using a Twirl template as done elsewhere in the example, are both good.

**Comment by bryce-anderson** [comment]
A look back in the [past](https://github.com/http4s/http4s/commit/528233db0970df1570956af2b1722a762e15ab32) shows that before we did twirl, the xml literal that formed the index had the mount prefix in the links. I don't know of a simple way to fix this otherwise.

**Comment by refried** [comment]
Oh yeah, I wondered at the time if there was an easy way to pass the mount
prefix to the twirl template, and whether that would be an appropriate fix.

**Comment by rossabaker** [comment]
I tried to skirt around that issue by making all the links relative.  It worked in the basic case, but when the redirect rerendered the index template from a Uri without a trailing slash, that's probably when the trouble began.

Request has a scriptName and pathInfo.  scriptName is supposed to reflect the mount prefix portion of the path, and pathInfo the remainder.  To make the template truly portable across mount prefixes:

1. I *think* Twirl template parameter groups can be implicit.  This may or may not help get a Request to the template.
2. I know the servlet backends set a scriptName (it's just the contextPath+servletPath).  Looking at the code, it looks like Blaze stores it in a UriTranslateRoot.translateRootKey.  This looks like trouble to me.

**Comment by bryce-anderson** [comment]
Thats an artifact of using ```UriTranslation``` for 'mounting'. Servlet stores similar information using the ```Request.Keys.PathInfoCaret```, but the actual uri is unaffected. I'll file a issue about the inconsistency of ```UriTranslation``` and the way servlet handles it.

**Comment by bryce-anderson** [comment]
I'm of the opinion we should patch this in whatever way we need to to get it working and fix associated problems in separate tickets: no need to keep an embarrassing broken demo while we solve the underlying problems. To that end, is there more issues to resolve than #180, and @refried, do you want to submit a PR or should we just handle it?

**Comment by rossabaker** [comment]
+1 to fixing what we have a fix for. The script name issue captured in the other ticket.

**Comment by refried** [comment]
@bryce-anderson Heh, I'm clumsy with git, and probably should've just let you guys handle it, but I guess it's good practice; thanks.

**Comment by bryce-anderson** [comment]
Should be fixed by #182 and #181. Thanks @refried.
